### PR TITLE
Remove ADID and ADIndex from CoreCLR

### PR DESCRIPTION
**Pull request by davidwrighton**
These types are legacy of our multi-appdomain support, and removing their tendrils simplifies many disparate portions of the codebase.

- Where not removed AppDomainIterators have been adjusted to work over in a single appdomain environment
- Finalization is now a simpler algorithm due to avoidance of multi-appdomain finalization behavior
- Domain Terminate paths have been removed
- Sync block no longer has any concept of appdomain index
- sync block header no longer has a concept of appdomain index
- Various infrastructure around entering/leaving appdomains is removed
  - As well as removing various loops which utilized said infrastructure

### CODE REVIEW

**Comment by janvorli** [comment]
@davidwrighton Awesome! I was always hoping to find a time slot to do this cleanup, so I am very happy that you beat me to it.

File: src/vm/stacksampler.h [hunk 1]
```diff
@@ -55,17 +54,15 @@ class StackSampler
     typedef MapSHash<MethodDesc*, CountInfo> CountInfoHash;
     typedef CountInfoHash::element_t CountInfoHashEntry;
 
-    typedef MapSHash<MethodDesc*, ADID> JitInfoHash;
```
Comment (janvorli) [inline src/vm/stacksampler.h]: That the only instance of `JitInfoHash`, the `m_jitInfo`, is not used for anything else than adding an entry into it. The only place where it was used for look up (`StackSampler::GetDomainId`) was removed. So it seems we can remove the typedef and the instance.

File: src/debug/ee/debugger.cpp [hunk 2]
```diff
@@ -14989,7 +14966,7 @@ HRESULT Debugger::IterateAppDomainsForPdbs()
 
     while (pADInfo)
     {
-        STRESS_LOG3(LF_CORDB, LL_INFO100, "Iterating over domain %#08x AD:%#08x %ls\n", pADInfo->m_pAppDomain->GetId().m_dwId, pADInfo->m_pAppDomain, pADInfo->m_szAppDomainName);
+        STRESS_LOG3(LF_CORDB, LL_INFO100, "Iterating over domain %#08x AD:%#08x %ls\n", DefaultADID, pADInfo->m_pAppDomain, pADInfo->m_szAppDomainName);
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/debug/ee/debugger.cpp]: Yes, that's part of the reason this is WIP. Most of the way through swapping to DefaultADID, I decided that was not a good plan, and started stripping them from STRESS_LOG commands. But I pushed it to repo to try and see how noisy the errors on other platforms would be.

File: src/debug/ee/dactable.cpp [hunk 3]
```diff
@@ -26,6 +22,10 @@
 #include "../../vm/gcenv.h"
 #include "../../vm/ecall.h"
 #include "../../vm/rcwwalker.h"
+// This header include will need to be rmeoved as part of GitHub#12170.
```
Comment (janvorli) [inline src/debug/ee/dactable.cpp]: A nit - rmeoved -> removed

**Comment by davidwrighton** [comment]
@Maoni0 Could you take a look at the finalization paths that I changed? I believe what I did looks good, but I think you're likely the expert there?

**Pull request merged**